specreduce/tracing.py

```diff
@@ -110,3 +113,164 @@ def __post_init__(self):
                 padding = np.ma.MaskedArray(np.ones(nx - nt) * self.trace[-1], mask=True)
                 self.trace = np.ma.hstack([self.trace, padding])
         self._bound_trace()
+
+
+@dataclass
+class KosmosTrace(Trace):
+    """
+    Trace the spectrum aperture in an image.
+
+    Chops image up in bins along the dispersion (wavelength) direction,
+    fits a Gaussian within each bin to determine the trace's spatial
+    center. Finally, draws a cubic spline through the bins to up-sample
+    trace along every pixel in the dispersion direction.
+
+    (The original version of this algorithm is sourced from James
+     Davenport's `kosmos` repository.)
+
+    Parameters
+    ----------
+    image : CCDData or array-like, required
+        The image over which to run the trace. Assumes cross-dispersion
+        (spatial) direction is axis 0 and dispersion (wavelength)
+        direction is axis 1.
+    bins : int, optional
+        The number of bins in the dispersion (wavelength) into which to
+        divide the image. Use fewer if KosmosTrace is having
+        difficulty, such as with faint targets. Minimum bin size is 4.
+        [default: 20]
+    guess : int, optional
+        A guess at the trace's location in the cross-dispersion
+        (spatial) direction. If set, overrides the normal max peak
+        finder. Good for tracing a fainter source if multiple traces
+        are present. [default: None]
+    window : int, optional
+        Fit the trace to a region with size `window * 2` aroudn the
+        guess position. Useful for tracing faint sources if multiple
+        traces are present, but potentially bad if the trace is
+        substantially bent or warped. [default: None]
+
+    Improvements Needed
+    -------------------
+    1) switch to astropy models for Gaussian (done)
+    2) return info about trace width (?)
+    3) add re-fit trace functionality (or break off into other method)
+    4) add other interpolation modes besides spline, maybe via
+        specutils.manipulation methods?
+    """
+    image: CCDData
+    bins: int = 20
+    guess: float = None
+    window: int = None
+    disp_axis = 1
+    
+    def __post_init__(self,
+                      # Saxis=0, Waxis=1, display=False
+                      ):
+        if self.bins < 4:
+            raise ValueError('bins must be >= 4')
+        if not isinstance(self.bins, int):
+            warnings.warn('TRACE: Converting bins to int')
+            self.bins = int(self.bins)
```

Comment:
I would switch these two so that the bins enforcement happens on rectified data
``` suggestion
        if not isinstance(self.bins, int):
            warnings.warn('TRACE: Converting bins to int')
            self.bins = int(self.bins)
        if self.bins < 4:
            raise ValueError('bins must be >= 4')
```